dist-kernel: drop support for <=sys-kernel/installkernel-gentoo-7

AndrewAmmerlaan opened the pull request:
**Aim:** Make it possible to use initrd generators other then dracut with our distribution kernels.

**Method:** Use the existing plugin system of systemd's `kernel-install` to allow customization of the kernel installation process with our distribution kernels.

**Functional changes:**
- `sys-apps/systemd-utils` can now optionally install `systemd-ukify` and its accompanying `kernel-install` plugin.
- When the user wants to use UKIs with our dist-kernels, it is no longer required to explicitly configure dracut for this with the `uefi=yes` option. Because the initrd generation is offloaded to `kernel-install` and the dracut plugin is smart enough to infer `uefi=yes` or `uefi=no` from the `kernel-install` `layout` setting.
- Customization and overriding of initrd generation is now possible if `sys-kernel/installkernel-systemd` is used.
- Installation of `kernel-install` by the `sys-apps/systemd(utils)` package is no longer coupled to the installation of `systemd-boot`. Though if USE=boot, then we still need USE=kernel-install. The long term aim here is provide `kernel-install` plugins for bootloaders other then `systemd-boot` (in so far as these don't exist already)
- `installkernel-systemd` now installs kernels with `--verbose`.

**Non-functional changes:**
-  `sys-kernel/installkernel-systemd-boot`  is renamed to `sys-kernel/installkernel-systemd`
- Minimum dracut version in `kernel-install.eclass` is bumped to 059-r4, this makes it possible to drop the workaroud we added earlier to fix the lack of UKI compatibility in dracut's `kernel-install` plugin.
- Workaround to skip dracut `kernel-install` plugin is dropped in `dist-kernel-utils.eclass`
- Initrd generation in the dist-kernels now occurs slightly later when  `sys-kernel/installkernel-systemd` is used.

**Still TODO:**
- Allow similar customization when `sys-kernel/installkernel-gentoo` is used instead. The most obvious route for this is to add initrd generation code to installkernel-gentoo, which then makes it possible to drop the initrd generation code from the eclasses completely.
- Or, once all bootloaders (and initrd generators) ship with proper systemd `kernel-install` plugins, simply drop `sys-kernel/installkernel-gentoo` and the initrd generation code in the eclass.

CC @gentoo/systemd for the changes to systemd(-utils)
CC @gentoo/dist-kernel @mgorny for the other changes

floppym commented on sys-apps/systemd-utils/systemd-utils-254.5-r2.ebuild:
```diff
@@ -425,14 +440,21 @@ multilib_src_install() {
+			doman man/kernel-install.8
+			# This may be empty depending on whether we also enable sd-boot/ukify
+			keepdir usr/lib/kernel/install.d
+			if [[ -n "$(ls -A src/kernel-install)" ]]; then
```
I would prefer a USE flag check instead of this `ls -A` hack.

AndrewAmmerlaan replied:
Yes this makes more sense indeed

AndrewAmmerlaan replied:
Actually, the systemd meson install adds some plugins already, if we simply install the same set of plugins by default in systemd-utils we avoid difference in behaviour between the two, and then we also don't need this whole condition after all.

floppym replied:
Sounds good.

AndrewAmmerlaan replied:
Fixed!

floppym reviewed (changes requested):
The systemd changes seem ok overall, just a few small changes.

floppym commented on sys-kernel/installkernel-systemd/installkernel-systemd-2-r4.ebuild:
```diff
@@ -0,0 +1,37 @@
+RDEPEND="
+	!sys-kernel/installkernel-gentoo
+	|| (
+		sys-apps/systemd[gnuefi(-)]
+		sys-apps/systemd[boot(-)]
+		sys-apps/systemd[kernel-install(-)]
+		sys-apps/systemd-utils[boot]
+		sys-apps/systemd-utils[kernel-install(-)]
```
Nevermind, I think I get it now.

AndrewAmmerlaan commented on sys-apps/systemd/systemd-9999.ebuild:
```diff
@@ -391,14 +395,16 @@ multilib_src_install_all() {
-		python_fix_shebang "${ED}"
-		secureboot_auto_sign
-	fi
+	use ukify && python_fix_shebang "${ED}"
```
I had this same thought earlier, but then I got confused when I noticed that systemd-utils also depends on the same python stuff, even though it doesn't install ukify. This might simply be a mistake though, since as far as I can see neither systemd-boot nor kernel-install does anything with python at all.

floppym replied:
Ah, could definitely be a mistake on my part.

AndrewAmmerlaan replied:
Fixed!

AndrewAmmerlaan commented on sys-apps/systemd-utils/systemd-utils-254.5-r2.ebuild:
```diff
@@ -425,14 +440,21 @@ multilib_src_install() {
+			dobin kernel-install
+			doman man/kernel-install.8
+			# This may be empty depending on whether we also enable sd-boot/ukify
+			keepdir usr/lib/kernel/install.d
```
Now that this directory is guaranteed not to be empty, we definitely don't need this.

AndrewAmmerlaan commented:
> The systemd changes seem ok overall, just a few small changes.

Had some troubles with the QA bot, but we're finally green again! If the systemd part is okay then I can merge this first, it is more or less independent of the changes to the eclass anyway.

As an added bonus we can now remove the mask of `sys-kernel/installkernel-systemd` on ppc64, since we now depend on `kernel-install` instead of on `systemd-boot` which avoids the dependency problem that caused the mask.

AndrewAmmerlaan commented:
Merged the systemd part, the dist-kernel eclass patches remain for review.

scardracs commented:
Hi @AndrewAmmerlaan!
I was talking with @floppym and wondering if it was possible to make a news item because not everyone care to make a research and asking 2/5/10 times why systemd has boot? (kernel-install) as blocker. This would be useful to state what actually kernel-install do (I know kernel-install do literally a kernel install, but the same is for ukify, which not everyone know what it actually is and how it works).
Would you think it would be useful?

AndrewAmmerlaan commented:
> Hi @AndrewAmmerlaan! I was talking with @floppym and wondering if it was possible to make a news item because not everyone care to make a research and asking 2/5/10 times on IRC why systemd has boot? (kernel-install) as blocker. This would be useful to state what actually kernel-install do (I know kernel-install do literally a kernel install, but the same is for ukify, which not everyone know what it actually is and how it works). Would you think it would be useful?

I'll see about writing some draft news item, but I think this should be about the whole thing (i.e. also the moving of the generation of the initrd).

AndrewAmmerlaan commented:
Merged the first two commits, and rebased. For the final commit it is probably best to first merge https://github.com/projg2/installkernel-gentoo/pull/10/

Here's some draft news item, I haven't posted this to the mailing list yet because it might still need changing if the patches here change after review: [2023-11-10-dist-kernel-initrd-gen.en.txt](https://github.com/gentoo/gentoo/files/13319187/2023-11-10-dist-kernel-initrd-gen.en.txt).

scardracs commented:
> > Hi @AndrewAmmerlaan! I was talking with @floppym and wondering if it was possible to make a news item because not everyone care to make a research and asking 2/5/10 times on IRC why systemd has boot? (kernel-install) as blocker. This would be useful to state what actually kernel-install do (I know kernel-install do literally a kernel install, but the same is for ukify, which not everyone know what it actually is and how it works). Would you think it would be useful?
> 
> I'll see about writing some draft news item, but I think this should be about the whole thing (i.e. also the moving of the generation of the initrd).

Yes, you know better how your update works and what would be useful to indicate on it. Ty :)

AndrewAmmerlaan commented:
@mgorny @ArsenArsen After the discussion we had on IRC I have reworked this to not change the existing behaviour of installkernel-gentoo (see https://github.com/projg2/installkernel-gentoo/pull/10). 

While debugging I found two small unrelated problems, which I fixed in commits 1 and 2 and which are now also on the mailing list. @mgorny please let me know if these two look okay, and then I'll merge them and rebase this PR again.

The third commit now contains the main eclass changes to accompany the installkernel-gentoo PR. The fourth commit serves as an example of how we can simplify the eclasses later when we drop support for the old versions <=7 of installkernel-gentoo. This is not for merging now, lets focus on the improved interaction with `kernel-install`, and the flexibility in initrd generation in commit 3 first.

I also thought about pre-generated initrds for gentoo-kernel-bin as mentioned on IRC. The way I think this could work (not something to implement now, but good to keep in mind for later), is to have some separate package (i.e. `sys-kernel/dracut-initrd-bin`) that contains the initrd that is already generated as a side product of the building of the kernels by binpkg-docker (https://github.com/projg2/binpkg-docker). This package is versioned with the kernel and installs the initrd to the usual location '/usr/src/linux-${KV}/arch/boot/${ARCH}/initrd`, gentoo-kernel-bin-x.y.z then DEPENDs on `=sys-kernel/dracut-initrd-bin-x.y.z` (we can either hard depend, or make it optional with some flag). 

With installkernel-systemd the initrd will be picked up at this location by `50-dracut.install` and it will not generate a new one. For installkernel-gentoo we can also easily make this work by making the `export KERNEL_INITRAMFS_GENERATOR` conditional to `has_version =sys-kernel/dracut-initrd-bin-${PV}`. This will ensure that with installkernel-gentoo as well the prebuilt version is used if it is installed.

We do have a fundamental incompatibility with a pre-generated initrd and `dist-kernel_reinstall_initramfs`, perhaps the best way to handle this is to warn or even error out if this function is called while the pre-generated initrd is installed. We only have one package using this function, so the impact of a hard error is small. Users who wish to use this package would not want to use the the pregenerated initrd anyway.

The pregenerated initrd might be something I can work on later, after we process this PR. In any case, after thinking about it, I personally don't see any reason why pregenerated initrds should be a reason to not move the responsibility of the initrd generations to installkernel*. `50-dracut.install` is already compatible with this, and with a single `if` we can easily do the same for installkernel-gentoo. The major problem is `dist-kernel_reinstall_initramfs`, but this is a problem regardless of where the generation of the initrd is taking place.

Let me know what you think!

AndrewAmmerlaan commented:
CI complaint happens because the fourth commit introduces deps on versions of installkernel-gentoo that do not exist yet, we can ignore it for now. The PR to installkernel-gentoo should be merged first anyway.

mgorny commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,43 +66,38 @@ dist-kernel_get_image_path() {
-	# initrd relatively to bzImage
-	ARCH=$(tc-arch-kernel) installkernel "${version}" "${image}" "${map}"
+	pushd ${kernel_dir} >/dev/null || die "Invalid kernel dir: ${kernel_dir}"
+	ARCH=$(tc-arch-kernel) emake install
```
NAK. That's hardly an improvement. You're relying on the `install` target having no side effects. Also, the kernel Makefiles hardcode paths, so fix that first.

AndrewAmmerlaan replied:
> NAK. That's hardly an improvement. You're relying on the `install` target having no side effects. Also, the kernel Makefiles hardcode paths, so fix that first.

Right, I hadn't thought about the hardcoded path, that is a good point. I've dropped this commit, leaving just the part that drops the code paths for installkernel-gentoo-7.  I don't really care about being able to set `INSTALLKERNEL` to some alternative script anyway, just thought it was a nice to have, but if it is going to require patching the kernel makefiles then it's not worth the effort I think.

mgorny commented on eclass/kernel-install.eclass:
```diff
@@ -743,7 +728,7 @@ kernel-install_pkg_prerm() {
 	debug-print-function ${FUNCNAME} "${@}"
 
-	if [[ -z ${ROOT} && ! ${KERNEL_IUSE_GENERIC_UKI} ]] && use initramfs; then
+	if [[ -z ${ROOT} && ! ${KERNEL_IUSE_GENERIC_UKI} ]]; then
```
I'm not exactly happy about retroactive change of behavior but I care about `USE=-initramfs` so little that *whatever*.

mgorny reviewed (approved):
LGTM
